base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -902,6 +922,105 @@ fn fetch_block<T: BlockchainBackend>(db: &T, height: u64) -> Result<HistoricalBl
     Ok(HistoricalBlock::new(block, tip_height - height + 1, spent))
 }
 
+fn fetch_block_with_kernel<T: BlockchainBackend>(
+    db: &T,
+    excess_sig: Signature,
+) -> Result<Option<HistoricalBlock>, ChainStorageError>
+{
+    let metadata = db.fetch_metadata()?;
+    let db_height = metadata.height_of_longest_chain.unwrap_or(0);
+    let horizon_height = metadata.horizon_block(db_height);
+    for i in (horizon_height..db_height).rev() {
+        let kernel_cp = fetch_checkpoint(db, MmrTree::Kernel, i)?;
+        let (kernel_hashes, _) = kernel_cp.into_parts();
+        let kernels = fetch_kernels(db, kernel_hashes)?;
+        for kernel in kernels {
+            if kernel.excess_sig == excess_sig {
+                return Ok(Some(fetch_block(db, i)?));
+            }
+        }
+    }
+    // data is not in the pruning horizon, lets check behind that but only if if there is a pruning horizon
+    if horizon_height > 0 {
+        let kernel_cp = fetch_checkpoint(db, MmrTree::Kernel, horizon_height - 1)?;
+        let (kernel_hashes, _) = kernel_cp.into_parts();
+        let kernels = fetch_kernels(db, kernel_hashes)?;
+        for kernel in kernels {
+            if kernel.excess_sig == excess_sig {
+                return Ok(None);
+            }
+        }
+    }
+    let err = format!("Block with kernel signature {}", excess_sig.get_signature().to_hex());
+    Err(ChainStorageError::ValueNotFound(err))
+}
+
+fn fetch_block_with_utxo<T: BlockchainBackend>(
+    db: &T,
+    commitment: Commitment,
+) -> Result<Option<HistoricalBlock>, ChainStorageError>
+{
+    let metadata = db.fetch_metadata()?;
+    let db_height = metadata.height_of_longest_chain.unwrap_or(0);
+    let horizon_height = metadata.horizon_block(db_height);
+    for i in (horizon_height..db_height).rev() {
+        let utxo_cp = fetch_checkpoint(db, MmrTree::Utxo, i)?;
+        let (utxo_hashes, _) = utxo_cp.into_parts();
+        let utxos = fetch_outputs(db, utxo_hashes)?;
+        for utxo in utxos.0 {
+            if utxo.commitment == commitment {
+                return Ok(Some(fetch_block(db, i)?));
+            }
+        }
+        for comm in utxos.1 {
+            if comm == commitment {
+                return Ok(Some(fetch_block(db, i)?));
+            }
+        }
+    }
+    // data is not in the pruning horizon, lets check behind that but only if if there is a pruning horizon
+    if horizon_height > 0 {
+        let utxo_cp = fetch_checkpoint(db, MmrTree::Utxo, horizon_height - 1)?;
+        let (utxo_hashes, _) = utxo_cp.into_parts();
+        let utxos = fetch_outputs(db, utxo_hashes)?;
+        for utxo in utxos.0 {
+            if utxo.commitment == commitment {
+                return Ok(None);
+            }
+        }
+        for comm in utxos.1 {
+            if comm == commitment {
+                return Ok(None);
+            }
+        }
```

Comment:
Destructuring the tuple is preferable

```suggestion
        let (outputs, spent) = fetch_outputs(db, utxo_hashes)?;
        for utxo in outputs {
            if utxo.commitment == commitment {
                return Ok(None);
            }
        }
        for commit in spent {
            if commit == commitment {
                return Ok(None);
            }
        }
```